storage/fcds/leveldb/leveldb.go

```diff
@@ -0,0 +1,176 @@
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package leveldb
+
+import (
+	"encoding/binary"
+
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/storage/fcds"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+var _ fcds.MetaStore = new(MetaStore)
+
+// MetaStore implements FCDS MetaStore with LevelDB
+// for persistence.
+type MetaStore struct {
+	db *leveldb.DB
+}
+
+// NewMetaStore returns new MetaStore at path.
+func NewMetaStore(path string) (s *MetaStore, err error) {
+	db, err := leveldb.OpenFile(path, &opt.Options{})
+	if err != nil {
+		return nil, err
+	}
+	return &MetaStore{
+		db: db,
+	}, err
+}
+
+// Get returns chunk meta information.
+func (s *MetaStore) Get(addr chunk.Address) (m *fcds.Meta, err error) {
+	data, err := s.db.Get(chunkKey(addr), nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return nil, chunk.ErrChunkNotFound
+		}
+		return nil, err
+	}
+	m = new(fcds.Meta)
+	if err := m.UnmarshalBinary(data); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+// Set adds a new chunk meta information for a shard.
+// Reclaimed flag denotes that the chunk is at the place of
+// already deleted chunk, not appended to the end of the file.
+func (s *MetaStore) Set(addr chunk.Address, shard uint8, reclaimed bool, m *fcds.Meta) (err error) {
+	batch := new(leveldb.Batch)
+	if reclaimed {
+		batch.Delete(freeKey(shard, m.Offset))
+	}
+	meta, err := m.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	batch.Put(chunkKey(addr), meta)
+	return s.db.Write(batch, nil)
+}
+
+// Remove removes chunk meta information from the shard.
+func (s *MetaStore) Remove(addr chunk.Address, shard uint8) (err error) {
+	m, err := s.Get(addr)
+	if err != nil {
+		return err
+	}
+	batch := new(leveldb.Batch)
+	batch.Put(freeKey(shard, m.Offset), nil)
+	batch.Delete(chunkKey(addr))
+	return s.db.Write(batch, nil)
+}
+
+// FreeOffset returns an offset that can be reclaimed by
+// another chunk. If the returned value is less then 0
+// there are no free offset at this shard.
+func (s *MetaStore) FreeOffset(shard uint8) (offset int64, err error) {
+	i := s.db.NewIterator(nil, nil)
+	defer i.Release()
+
+	i.Seek([]byte{freePrefix, shard})
+	key := i.Key()
+	if key == nil || key[0] != freePrefix || key[1] != shard {
```

Comment:
Keeping track of a count is additional operation on leveldb involving seek/read/write in Set and Remove. The case where there are no free slots, Store.free is optimising that in memory to avoid MetaStore.FreeOffest call which would involve disk i/o seek/read in this implementation.

Reply:
wait you do not need to keep the count on disk. you get slotcount from the shardsize and get freecount with iteration on free slots once at startup. After that it is just in memory, no?

Reply:
Iteration on start may take arbitrary long time. I do not think that it is needed.